src/game/engine/smudge.cpp

```diff
@@ -0,0 +1,241 @@
+                    if (type->Is_Crater() && cptr->Has_Crater()) {
+                        cptr->Set_Smudge_Frame(cptr->Get_Smudge_Frame() + 1);
+                        cptr->Set_Smudge_Frame(std::min(cptr->Get_Smudge_Frame(), (uint8_t)4));
+                    }
+                    if (cptr->Get_Smudge() == SMUDGE_NONE) {
+                        cptr->Set_Smudge(type->Is_Crater() ? (SmudgeType)cptr->Spot_Index(m_Coord) : type->What_Type());
+                        cptr->Set_Smudge_Frame(0);
+                    }
+                }
+                cptr->Redraw_Objects();
+            }
+        }
+    }
+    Map.Overlap_Up(Get_Cell(), this);
+    m_IsDown = false;
+    m_InLimbo = true;
+    delete this;
+    return true;
+}
+
+/**
+ *
+ *
+ */
+void SmudgeClass::Disown(cell_t cell)
+{
+    const SmudgeTypeClass *type = &Class_Of();
+    if (type->Is_Bib()) {
+        for (int i = 0; i < type->Get_Width(); ++i) {
+            for (int j = 0; j < type->Get_Height(); ++j) {
+                cell_t calc_cell = (j * CELL_MAX) + i + cell;
+                CellClass *cptr = &Map[calc_cell];
+                if (cptr->Get_Overlay() == OVERLAY_NONE || !cptr->Has_Wall()) {
+                    cptr->Set_Smudge(SMUDGE_NONE);
+                    cptr->Set_Smudge_Frame(0);
+                    if (!cptr->Has_Flag()) {
+                        cptr->Set_Owner(HOUSES_NONE);
+                    }
+                    cptr->Redraw_Objects();
+                }
+            }
+        }
+    }
+}
+
+/**
+ *
+ *
+ */
+void SmudgeClass::Read_INI(GameINIClass &ini)
+{
+    char valuebuf[128];
+
+    for (int i = 0; i < ini.Entry_Count("SMUDGE"); ++i) {
+        ini.Get_String("SMUDGE", ini.Get_Entry("SMUDGE", i), nullptr, valuebuf, sizeof(valuebuf));
+        SmudgeType smudge = SmudgeTypeClass::From_Name(strtok(valuebuf, ","));
+        if (smudge != SMUDGE_NONE) {
+            char *cell_str = strtok(nullptr, ",");
+            if (cell_str != nullptr) {
+                cell_t cellnum = atoi(cell_str);
+
+                int frame = 0;
+                char *frame_str = strtok(nullptr, ",");
+                if (frame_str != nullptr) {
+                    frame = atoi(frame_str);
+                }
+
+                new SmudgeClass(smudge, Cell_To_Coord(cellnum));
+                CellClass &cell = Map[cellnum];
+                if (cell.Get_Smudge() == smudge && frame > 0) {
+                    cell.Set_Smudge_Frame(frame);
+                }
+            }
+        }
+    }
+}
+
+/**
+ *
+ *
+ */
+void SmudgeClass::Write_INI(GameINIClass &ini)
+{
+    char num[32];
+    char entry[128];
+
+    ini.Clear("SMUDGE");
+    for (int i = 0; i < MAP_MAX_AREA; ++i) {
+        CellClass *cptr = &Map[i];
+        if (cptr->Get_Smudge() != SMUDGE_NONE) {
+            SmudgeTypeClass *stptr = SmudgeTypeClass::As_Pointer(cptr->Get_Smudge());
+            if (stptr != nullptr && !stptr->Is_Bib()) {
+                std::snprintf(num, sizeof(num), "%d", i);
```

Comment:
Why are you using std:: here, but not further up the file on strtok? Should be including <cstring> and <cstdio> and have `using std::snprintf;` and `using std::strtok;` at the top of the file for C library functions.